diff --git a/vsim/Makefrag-verilog b/vsim/Makefrag-verilog
index 66034bc..e2906a2 100644
--- a/vsim/Makefrag-verilog
+++ b/vsim/Makefrag-verilog
@@ -7,7 +7,11 @@ $(generated_dir)/$(MODEL).$(CONFIG).v $(generated_dir)/$(MODEL).$(CONFIG).d $(ge
 	cd $(base_dir) && mkdir -p $(generated_dir) && $(SBT) "run $(CHISEL_ARGS) --configDump --noInlineMem"
 	cd $(generated_dir) && \
 	if [ -a $(MODEL).$(CONFIG).conf ]; then \
-	  $(mem_gen) $(generated_dir)/$(MODEL).$(CONFIG).conf >> $(generated_dir)/$(MODEL).$(CONFIG).v; \
+	  if [[ $(CONFIG) == *XFiles* ]]; then \
+	    $(mem_gen) -c $(generated_dir)/$(MODEL).$(CONFIG).conf -I ../../xfiles-dana/src/main/verilog -I ../../xfiles-dana/submodules/verilog/src -v sram_r1_w1_rw0.v >> $(generated_dir)/$(MODEL).$(CONFIG).v; \
+	  else \
+	    $(mem_gen) -c $(generated_dir)/$(MODEL).$(CONFIG).conf >> $(generated_dir)/$(MODEL).$(CONFIG).v; \
+	  fi; \
 	fi
 
 $(generated_dir)/memdessertMemDessert.$(CONFIG).v $(generated_dir)/memdessertMemDessert.$(CONFIG).d: $(base_dir)/$(src_path)/*.scala $(base_dir)/uncore/$(src_path)/*.scala
