diff --git a/Makefile.in b/Makefile.in
index 8babada..9cec4ed 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -37,6 +37,8 @@ src_dir      := @srcdir@
 scripts_dir  := $(src_dir)/scripts
 bbl_payload  := @BBL_PAYLOAD@
 
+ARCH ?= RV64IMAFDXcustom
+
 # If the version information is not in the configure script, then we
 # assume that we are in a working directory. We use the vcs-version.sh
 # script in the scripts directory to generate an appropriate version
@@ -72,11 +74,12 @@ install_exes_dir := $(INSTALLDIR)/bin
 sprojs         := @subprojects@
 sprojs_enabled := @subprojects_enabled@
 
-sprojs_include := -I. $(addprefix -I$(src_dir)/, $(sprojs_enabled))
+sprojs_include := -I. $(addprefix -I$(src_dir)/, $(sprojs_enabled)) \
+	-I$(src_dir)/../../xfiles-dana
 VPATH := $(addprefix $(src_dir)/, $(sprojs_enabled))
 
 #-------------------------------------------------------------------------
-# Programs and flags 
+# Programs and flags
 #-------------------------------------------------------------------------
 
 # C++ compiler
@@ -84,7 +87,7 @@ VPATH := $(addprefix $(src_dir)/, $(sprojs_enabled))
 #  - CXXFLAGS : flags for C++ compiler (eg. -Wall,-g,-O3)
 
 CC            := @CC@
-CFLAGS        := @CFLAGS@ $(CFLAGS) -DBBL_PAYLOAD=\"$(bbl_payload)\"
+CFLAGS        := @CFLAGS@ $(CFLAGS) -DBBL_PAYLOAD=\"$(bbl_payload)\" -march=$(ARCH)
 COMPILE       := $(CC) -MMD -MP $(CFLAGS) \
                  $(sprojs_include)
 # Linker
@@ -130,7 +133,7 @@ dist_junk += $(sprojs_mk)
 #-------------------------------------------------------------------------
 # This function is used by the subproject template to reverse the list
 # of dependencies. It uses recursion to perform the reversal.
-# 
+#
 # Arguments:
 #  $(1)   : space separated input list
 #  retval : input list in reverse order
@@ -152,7 +155,7 @@ endef
 # subprojects defining a certain set of make variables which are all
 # prefixed with the subproject name. Since subproject names can have
 # dashes in them (and the make variables are assumed to only use
-# underscores) the template takes two arguments - one with the regular 
+# underscores) the template takes two arguments - one with the regular
 # subproject name and one with dashes replaced with underscores.
 #
 # Arguments:
@@ -282,7 +285,7 @@ clean-$(1) :
 
 libs += lib$(1).a
 objs += $$($(2)_c_objs)
-srcs += $$(addprefix $(src_dir)/$(1)/, $$($(2)_c_srcs)) 
+srcs += $$(addprefix $(src_dir)/$(1)/, $$($(2)_c_srcs))
 hdrs += $$(addprefix $(src_dir)/$(1)/, $$($(2)_hdrs))
 junk += $$($(2)_junk)
 deps += $$($(2)_c_deps)
@@ -373,7 +376,7 @@ dist_junk += config.status config.h Makefile config.log
 # Distribution
 #-------------------------------------------------------------------------
 # The distribution tarball is named project-ver.tar.gz and it includes
-# both enabled and disabled subprojects. 
+# both enabled and disabled subprojects.
 
 dist_files = \
   $(sprojs) \
diff --git a/configure b/configure
index 6386de0..833c854 100755
--- a/configure
+++ b/configure
@@ -3977,7 +3977,7 @@ fi
 
 CFLAGS="-Wall -Werror -D__NO_INLINE__ -mcmodel=medany -O2 -std=gnu99 -Wno-unused -Wno-attributes -fno-delete-null-pointer-checks"
 
-LIBS="-lgcc"
+LIBS="-lgcc -L../../../xfiles-dana/usr/lib-rv-newlib -lxfiles-supervisor"
 
 
 #-------------------------------------------------------------------------
diff --git a/configure.ac b/configure.ac
index 6253228..23b1659 100644
--- a/configure.ac
+++ b/configure.ac
@@ -72,7 +72,7 @@ AC_HEADER_STDC
 #-------------------------------------------------------------------------
 
 AC_SUBST([CFLAGS],  ["-Wall -Werror -D__NO_INLINE__ -mcmodel=medany -O2 -std=gnu99 -Wno-unused -Wno-attributes -fno-delete-null-pointer-checks"])
-AC_SUBST([LIBS],  ["-lgcc"])
+AC_SUBST([LIBS],  ["-lgcc -L../../../xfiles-dana/usr/lib-rv-newlib -lxfiles-supervisor"])
 
 #-------------------------------------------------------------------------
 # MCPPBS subproject list
diff --git a/machine/minit.c b/machine/minit.c
index 055fdb3..aa03338 100644
--- a/machine/minit.c
+++ b/machine/minit.c
@@ -15,6 +15,7 @@ static void mstatus_init()
   uintptr_t ms = 0;
   ms = INSERT_FIELD(ms, MSTATUS_VM, VM_CHOICE);
   ms = INSERT_FIELD(ms, MSTATUS_FS, 1);
+  ms = INSERT_FIELD(ms, MSTATUS_XS, 1);
   write_csr(mstatus, ms);
 
   // Make sure the hart actually supports the VM mode we want
diff --git a/pk/syscall.c b/pk/syscall.c
index de11a0a..e2b5b27 100644
--- a/pk/syscall.c
+++ b/pk/syscall.c
@@ -9,6 +9,8 @@
 #include <string.h>
 #include <errno.h>
 
+#include "src/main/c/xfiles-supervisor.h"
+
 typedef long (*syscall_t)(long, long, long, long, long, long, long);
 
 #define CLOCK_FREQ 1000000000
@@ -380,7 +382,7 @@ int sys_times(long* loc)
   loc[1] = 0;
   loc[2] = 0;
   loc[3] = 0;
-  
+
   return 0;
 }
 
@@ -389,7 +391,7 @@ int sys_gettimeofday(long* loc)
   uintptr_t t = rdcycle();
   loc[0] = t / CLOCK_FREQ;
   loc[1] = (t % CLOCK_FREQ) / (CLOCK_FREQ / 1000000);
-  
+
   return 0;
 }
 
@@ -421,6 +423,17 @@ static int sys_stub_nosys()
   return -ENOSYS;
 }
 
+int sys_xfiles_dana_set_asid(asid_type asid)
+{
+  return set_asid(asid);
+}
+
+int sys_xfiles_dana_set_antp(asid_nnid_table_entry * antp, size_t size)
+{
+  return set_antp(antp, size);
+}
+
+
 long do_syscall(long a0, long a1, long a2, long a3, long a4, long a5, unsigned long n)
 {
   const static void* syscall_table[] = {
@@ -466,6 +479,8 @@ long do_syscall(long a0, long a1, long a2, long a3, long a4, long a5, unsigned l
     [SYS_getrusage] = sys_stub_nosys,
     [SYS_getrlimit] = sys_stub_nosys,
     [SYS_setrlimit] = sys_stub_nosys,
+    [SYS_xfiles_dana_set_asid] = sys_xfiles_dana_set_asid,
+    [SYS_xfiles_dana_set_antp] = sys_xfiles_dana_set_antp,
   };
 
   const static void* old_syscall_table[] = {
diff --git a/pk/syscall.h b/pk/syscall.h
index c1f3d8a..ba69beb 100644
--- a/pk/syscall.h
+++ b/pk/syscall.h
@@ -50,6 +50,9 @@
 #define SYS_getrusage 165
 #define SYS_clock_gettime 113
 
+#define SYS_xfiles_dana_set_asid 512
+#define SYS_xfiles_dana_set_antp 513
+
 #define OLD_SYSCALL_THRESHOLD 1024
 #define SYS_open 1024
 #define SYS_link 1025
